proposals/shared-everything-threads/Overview.md

```diff
@@ -702,6 +702,37 @@ models. For more discussion: [How should toolchains apply `shared`?][toolchain-s
 
 [toolchain-shared-discussion]: https://github.com/abrown/thread-spawn/discussions/5
 
+### How will GCs and engines need to change?
+
+GCs that historically assume that there is only a single mutator thread will have to newly be able
+to handle heaps with multiple mutators running in parallel. Since sharedness is part of static
+types, engines can know at allocation or access time when they need to do something different for
+shared objects, so it is possible for them to add a separate concept of shared heap on top of their
+existing single-threaded heaps. This might be a reasonable intermediate architecture, even for
+engines that plan on eventually switching to having shared heaps as their only primitive.
+
+For engines that choose to have both shared and unshared heaps, references from unshared objects to
+shared objects will have to be treated as roots when the garbage collection runs on the shared heap.
+
+References from shared objects to unshared are normally disallowed, but we could choose to support
+them in limited cases to make TLS or thread-bound data more ergonomic. If we do, there are varying
+guarantees we could choose to make about how the GC treats these references.
+
+The weakest semantics would be that shared-to-unshared references never suffice to keep the unshared
+object alive. Users would have to root unshared objects for the duration of their intended
+lifetimes, so this is no more expressive than simply disallowing shared-to-unshared references.
+
+The strongest semantics would be that shared-to-unshared references behave exactly like normal
+references with respect to garbage collection. This requires the garbage collector to have
+visibility over the shared and unshared heaps from all threads so it can find unreachable cross-heap
+cycles.
+
+An intermediate semantics would be the same as the strong semantics, but without the guarantee that
+cross-heap cycles will be collected. This is as expressive as the weak semantics augmented with
+uncollectible shared-key finalization registries and allows the garbage collector to use a snapshot
```

Comment:
Oh, I meant that the hypothetical finalization registries are uncollectible, not that their keys would be uncollectible. Will clarify.

Do all the APIs that hold things weakly exhibit the behavior of not collecting weak cycles? I thought `FinalizationRegistry` was special because it can be used to emulate `WeakMap` _except_ for the collecting of cycles.

Reply:
> Do all the APIs that hold things weakly exhibit the behavior of not collecting weak cycles? I thought FinalizationRegistry was special because it can be used to emulate WeakMap except for the collecting of cycles.

Oh right, yes, that point is true. But I'd say it's that `WeakMap` (ephemeron semantics) that's special, not `FinalizationRegistry` and `WeakRef`. Ephemerons can collect cycles because they "hook into" the marking phase of the GC. Normal weak refs like `FinalizationRegistry` and `WeakRef` only allow observation of being collectible, which is _after_ marking.